src/os_linux/os_futex.c

```diff
@@ -0,0 +1,51 @@
+/*-
+ * Copyright (c) 2024-present MongoDB, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+#include <linux/futex.h>
+#include <sys/syscall.h>
+
+/*
+ * __wt_futex_wait --
+ *     Wait on the futex. The timeout is in microseconds and MUST be greater than zero.
+ */
+int
+__wt_futex_wait(
+  volatile WT_FUTEX_WORD *addr, WT_FUTEX_WORD expected, time_t usec, WT_FUTEX_WORD *wake_valp)
+{
+    struct timespec timeout;
+    long sysret;
+
+    WT_ASSERT(NULL, usec > 0);
+
+    sysret = syscall(SYS_futex, addr, FUTEX_WAIT_PRIVATE, expected,
+      __wt_usec_as_timespec(usec, &timeout), NULL, 0);
+    if (sysret == 0)
+        *wake_valp = __atomic_load_n(addr, __ATOMIC_SEQ_CST);
+
+    return ((int)sysret);
+}
+
+/*
+ * __wt_futex_wake --
+ *     Wake the futex.
+ */
+int
+__wt_futex_wake(volatile WT_FUTEX_WORD *addr, WT_FUTEX_WAKE wake, WT_FUTEX_WORD wake_val)
+{
+    long sysret;
+    int wake_op;
+
+    WT_ASSERT(NULL, wake == WT_FUTEX_WAKE_ONE || wake == WT_FUTEX_WAKE_ALL);
+
+    wake_op = (wake == WT_FUTEX_WAKE_ALL) ? INT_MAX : 1;
+    __atomic_store_n(addr, wake_val, __ATOMIC_SEQ_CST);
```

Comment:
This is probably an Andrew question, but at the risk of kicking a hornet's nest - are we at the point of directly using C atomics in WiredTiger? I thought we were meant to go through the wrappers.

Reply:
The wrappers for the wt atomic load store have relaxed semantics. I'm not sure that is strict enough for this use case (it might be) so I used the strictest ordering. As PM-3221 is still ongoing I did not want to start monkeying with the api.

Reply:
Fair enough. Thanks for explaining that, I'm not following that work as closely as I probably should be.